feat: switch to a per-device metrics model for vsock and rng devices

aerosouund opened the pull request:
## Changes

Building on the original [PR](https://github.com/firecracker-microvm/firecracker/pull/5196) and addressing the reviewer comments.
With the following remarks:

- Vsock device metrics are keyed in the btreemap by the guest CID which is hardcoded for test instance to be 52. if the desire is to run the tests in a given file, it may be needed to not hardcode it
- Entropy device metrics use a random generated UUID as the key of the map, because entropy devices don't have a unique identifier in the system that distinguishes them from one another (if one exists, let me know about it)
 
## Reason

Closes https://github.com/firecracker-microvm/firecracker/issues/4709

## License Acceptance

By submitting this pull request, I confirm that my contribution is made under
the terms of the Apache 2.0 license. For more information on following Developer
Certificate of Origin and signing off your commits, please check
[`CONTRIBUTING.md`][3].

## PR Checklist

- [x] I have read and understand [CONTRIBUTING.md][3].
- [x] I have run `tools/devtool checkbuild --all` to verify that the PR passes
  build checks on all supported architectures.
- [x] I have run `tools/devtool checkstyle` to verify that the PR passes the
  automated style checks.
- [x] I have described what is done in these changes, why they are needed, and
  how they are solving the problem in a clear and encompassing way.
- [x] I have updated any relevant documentation (both in code and in the docs)
  in the PR.
- [ ] I have mentioned all user-facing changes in `CHANGELOG.md`.
- [ ] If a specific issue led to this PR, this PR closes the issue.
- [ ] When making API changes, I have followed the
  [Runbook for Firecracker API changes][2].
- [ ] I have tested all new and changed functionalities in unit tests and/or
  integration tests.
- [ ] I have linked an issue to every new `TODO`.

______________________________________________________________________

- [x] This functionality cannot be added in [`rust-vmm`][1].

[1]: https://github.com/rust-vmm
[2]: https://github.com/firecracker-microvm/firecracker/blob/main/docs/api-change-runbook.md
[3]: https://github.com/firecracker-microvm/firecracker/blob/main/CONTRIBUTING.md

Manciukic commented on src/vmm/src/devices/virtio/rng/metrics.rs:
```diff
@@ -33,23 +33,51 @@
+});
+
+#[derive(Debug, Serialize, Default)]
+pub struct EntropyDeviceMetrics {
```
I don't think we should copy the whole "PerDevice" pattern for devices that only support a single instance.
The main idea of the change is correct: decoupling the device object from the global METRICS so unit tests are independent. However, we can probably simplify this part to just hold a single `Arc` in the `RwLock`, with the `Arc` being created by the device (and not the other way around like in net and block, as that will make it being reused).

Manciukic commented on src/vmm/src/devices/virtio/rng/device.rs:
```diff
@@ -488,34 +512,45 @@ mod tests {
+            &dev.metrics.entropy_event_fails,
             1,
             dev.process_rate_limiter_event()
         );
```
We should not need this macro hack anymore, and we could simplify this code block back to just checking the metric is now 1.

```
dev.process_rate_limiter_event();
assert_eq!(dev.metrics.entropy_event_fails, 1);
```

Manciukic replied:
yes, if we can simplify it as above, it would be ideal

```diff
-        check_metric_after_block!(
-            &METRICS.entropy_event_fails,
-            1,
-            dev.process_rate_limiter_event()
-        );
+        dev.process_rate_limiter_event();
+        assert_eq!(dev.metrics.entropy_event_fails, 1);
```

Maybe it's best to split the two changes in two commits:
 - first commit to use the local reference as you did here
 - later commit to simplify to a simple assertion

Manciukic reviewed (commented):
Thanks for the contibution! 

I only had a quick look, but I think we can simplify the logic for devices that only allow a single instance. Maybe a simple `RwLock<Arc<Metrics>>` would do the job.

Once we align on the solution for single-entity devices, we should also tackle all other devices using global metrics (like mem and pmem which we recently added), and simplify their unit tests as well.

Please keep different devices in different commits for clarity (and feel free to tackle one at a time if you prefer).

I'm also not a big fan of the current per-device metrics for net and block sharing the same metrics Arc if the ID happens to be the same, which defeats the purpose of not sharing metrics. As in real life we can't have devices with the same ID, I'd rather have the device create the object and replace whatever is in the hashmap rather than the other way around.

Manciukic commented:
> single instance devices create their arc during initialization and store it directly in the RwLock and multi instance devices also create their own metrics but store it in the hashmap themselves rather than call alloc to give them a metrics instance ?

Yes, that's what I meant. It seems it would simplify your implementation for the rng device quite much

Manciukic commented on src/vmm/src/devices/virtio/net/tap.rs:
```diff
@@ -241,10 +244,16 @@ pub mod tests {
+            Regex::new(r"^tap\d+$").unwrap().is_match(tap_name_str),
+            "Generated tap name '{}' does not match expected pattern",
+            tap_name_str
+        );
```
can you split it to a different commit with a clear subject and message?

Manciukic commented on src/vmm/src/devices/virtio/rng/device.rs:
```diff
@@ -446,29 +455,36 @@ mod tests {
+        assert_eq!(th.device().metrics.entropy_event_fails.count(), 1);
+        assert_eq!(th.device().metrics.entropy_event_count.count(), 1);
+        assert_eq!(th.device().metrics.entropy_bytes.count(), entropy_bytes);
+        assert_eq!(th.device().metrics.host_rng_fails.count(), host_rng_fails);
```
these should just be `0`.

Manciukic commented on src/vmm/src/devices/virtio/rng/metrics.rs:
```diff
@@ -84,17 +85,18 @@ impl EntropyDeviceMetrics {
+        assert!(METRICS.get().unwrap().activate_fails.count() >= 1);
+        assert!(METRICS.get().unwrap().entropy_bytes.count() >= 10);
+        assert_eq!(METRICS.get().unwrap().host_rng_fails.count(), 5);
     }
```
let's also test the serialization works (the previous test was at least checking it wasn't crashing)

Manciukic reviewed (commented):
I only skimmed through it but I think the approach is the right one. A few things to iron out:
 - We can improve all those unit tests that are reading the old values, doing an operation, and reading the new value and comparing. This was needed because of the global metrics. They should all now become simple assertions on the value of the metric after the operation. We don't need that macro anymore as well
 - We should split changes to different devices in different commits to make it simpler to review. The same for unrelated changes and the improvement to tests
 - Please ensure the PR passes the formatter and buildchecks by running the devtool (I left instructions in one of the comments)
 - We should do the same for all other devices to enable testing in parallel. If you have the time that'd be highly appreciated!

Thanks again for the contribution and sorry for the late reply!

aerosouund commented:
@Manciukic 
Thanks for taking the time to review it. Will address your comments along with the CI failures shortly

aerosouund committed:
refactor: Make the balloon device create its own metrics instance

- replace check_metric_after_block with traditional asserts
- replace access to METRICS for incrementing with self.metrics
- make report_balloon_event_fail take an arc of BalloonDeviceMetrics
  as an argument to avoid having to reset the METRICS instance after
  every test. given that each device will instantiate the metrics
  instance during the start of every test
- remove the comparison of the empty local metrics instance with the
  flushed global instance in test_balloon_dev_metrics

Signed-off-by: aerosouund <aerosound161@gmail.com>

aerosouund committed:
refactor: Make the block device create its own metrics instance

- replace check_metric_after_block with traditional asserts
- replace access to METRICS for incrementing with self.metrics
- change values in tests with asserts that check the total not the delta

Signed-off-by: aerosouund <aerosound161@gmail.com>

aerosouund committed:
refactor: Make the pmem devices create their own metric instances

- delete the PmemMetricsPerDevice type
- turn METRICS into an rwlock of the metrics btreemap directly
- make a device create its own metrics instance then add it to
  the METRICS map which is keyed on the device ids
- delete test_max_pmem_dev_metrics test because if the we remove
  reliance on the METRICS instance, it just becomes a test that checks
  counters incremenet for a group of metrics instance in a vector.
  meaning that it adds nothing over test_single_pmem_dev_metrics

Signed-off-by: aerosouund <aerosound161@gmail.com>

aerosouund committed:
refactor: Make the rng device create its own metrics instance

- replace check_metric_after_block with traditional asserts
- replace access to METRICS for incrementing with self.metrics
- change some test values to assert the absolute value not the delta
- depend on a local instance in test_rng_dev_metrics

Signed-off-by: aerosouund <aerosound161@gmail.com>

aerosouund committed:
refactor: Make the net device own its metrics creation

- delete the NetMetricsPerDevice type and make METRICS
  an rwlock of a btreemap directly, keyed on the net device id field.
- during net device initialization push the an arc of device metrics to
  the metrics map.
- remove reliance on the global metrics instance in
  test_max_net_dev_metrics and test_single_net_dev_metrics by running
  tests against local instances

Signed-off-by: aerosouund <aerosound161@gmail.com>

make the net dev own

Signed-off-by: aerosouund <aerosound161@gmail.com>

aerosouund committed:
refactor: Make the mem device create its own metrics instance

- replace access to METRICS for incrementing with self.metrics
- add derive default to VirtioMemDeviceMetrics
- make the METRICS variable a oncelock of an arc of
  VirtioMemDeviceMetrics and initialize the lock during device init

Signed-off-by: aerosouund <aerosound161@gmail.com>

aerosouund committed:
refactor: Make the vsock muxer create its own metrics instance

- turn the METRICS static variable to a btreemap of metrics arc's keyed
  on guest cid.
- the vsock multiplexer is the central piece and the entity that gets
  created first among the different vsock structs. make it create an
  arc of VsockDeviceMetrics and add that to the METRICS map.
- add aggregate() function to VsockDeviceMetrics.
- add unit tests in metrics.rs for aggregation and formatting.
- remove check_metrics_after_block use and replace it with traditional
  asserts. update the assertion values in select tests to make it
  reflect the real value not the delta
- make the connection type either fetch the metrics map entry or create
  its own if not found (test scenario)

Signed-off-by: aerosouund <aerosound161@gmail.com>

aerosouund committed:
refactor: replace check_metric_after_block in msix tests

- it has been replaced by pure asserts

Signed-off-by: aerosouund <aerosound161@gmail.com>

aerosouund committed:
cleanup: delete the check_metric_after_block macro

- its not in use anywhere after the previous commits

Signed-off-by: aerosouund <aerosound161@gmail.com>

aerosouund committed:
refactor: change vsock device init args

- make the vsock device struct take an option of the metric instance
- since the vsock muxer is what creates the metric instance and stores
  it in a map and the device needing to increment to it too, make it a
  parameter of the device initialization as an Option. because in the
  test cases you may want to supply a dummy backend so it won't have a
  valid metrics instance. if the option is None, the device creates its
  own metrics instance. but doesn't add it to the map to avoid
  overriding a muxer's instance and having to handle an existing
  instance in the METRICS map during muxer init.

Signed-off-by: aerosouund <aerosound161@gmail.com>

aerosouund commented on src/vmm/src/devices/virtio/balloon/metrics.rs:
```diff
@@ -35,22 +35,26 @@
-    seq.serialize_entry("balloon", &METRICS)?;
+    let metrics = METRICS
+        .get()
+        .expect("balloon: metrics instance not intialized");
```
I am indeed observing that this path gets triggered in test failures, though i'd like to know how it gets triggered and why is the effect of this is a long halt rather than an immediate exit
```
2026-03-04T22:43:29.756432256 [a76a7552-015d-4c6f-a242-737a7dcd0bd7:main] Firecracker panicked at /firecracker/src/vmm/src/devices/virtio/balloon/metrics.rs:50:10:
balloon: metrics instance not intialized
2026-03-04T22:43:29.757522359 [a76a7552-015d-4c6f-a242-737a7dcd0bd7:fc_api] The API server received a Get request on "/".
```

aerosouund replied:
@Manciukic 
got it, the reason for the timeout is because we flush metrics right away on the start of the vm when we call `vm.spawn()` in the python test

```rust
    let result = build_result.and_then(|vmm| {
        firecracker_metrics
            .lock()
            .expect("Poisoned lock")
            .start(super::metrics::WRITE_METRICS_PERIOD_MS); // here

        ApiServerAdapter::run_microvm(api_event_fd, from_api, to_api, vmm, &mut event_manager)
    });
```

this results in the following:

```rust
    pub(crate) fn start(&mut self, interval_ms: u64) {
        // Arm the log write timer.
        let duration = Duration::from_millis(interval_ms);
        self.write_metrics_event_fd.arm(duration, Some(duration));

        // Write the metrics straight away to check the process startup time.
        self.write_metrics();
    }

    fn write_metrics(&mut self) {
        if let Err(err) = METRICS.write() {
            METRICS.logger.missed_metrics_count.inc();
            error!("Failed to write metrics: {}", err);
        }

        #[cfg(test)]
        {
            self.flush_counter += 1;
        }
    }
```

and the call to `METRICS.write()` will trigger a flush to the metrics of all devices through the macro call `create_serialize_proxy!`. this will cause a panic and all subsequent HTTP api calls will hang.
do you think there should be a way to handle this more gracefully through having timeout settings on the python api calls ?

aerosouund committed:
refactor: serialize an empty instance in if METRICS is empty

- change affects balloon, rng and mem devices

Signed-off-by: aerosouund <aerosound161@gmail.com>

aerosouund commented on src/vmm/src/devices/virtio/balloon/metrics.rs:
```diff
@@ -35,22 +35,25 @@
-    seq.serialize_entry("balloon", &METRICS)?;
+    if let Some(metrics) = METRICS.get() {
+        seq.serialize_entry("balloon", &metrics)?;
+    };
```
@Manciukic 
apologies for the confusion. done, and fixed the failing test.
```
integration_tests/functional/test_metrics.py::test_flush_metrics[vmlinux-5.10.245-PCI_ON] PASSED [ 16%]
integration_tests/functional/test_metrics.py::test_flush_metrics[vmlinux-5.10.245-PCI_OFF] PASSED [ 33%]
integration_tests/functional/test_metrics.py::test_net_metrics[vmlinux-5.10.245-PCI_ON] PASSED [ 50%]
integration_tests/functional/test_metrics.py::test_net_metrics[vmlinux-5.10.245-PCI_OFF] PASSED [ 66%]
integration_tests/functional/test_metrics.py::test_block_metrics[vmlinux-5.10.245-PCI_ON] PASSED [ 83%]
integration_tests/functional/test_metrics.py::test_block_metrics[vmlinux-5.10.245-PCI_OFF] PASSED [100%]
```

aerosouund committed:
fix: adjust vsock serialziation

- vsock metrics instances were being serialized under the vm cid. not
  the word "vsock" as the schema expects

Signed-off-by: aerosouund <aerosound161@gmail.com>

aerosouund commented:
@Manciukic 
sorry for the delay. fixed the merge conflicts and the style checks

aerosouund commented:
@Manciukic 
fixed the new merge conflicts again. its worth noting also that the last CI check you ran did pass. there was only one flake on of the lanes for the functional tests

aerosouund commented:
@Manciukic 
fixed the merge conflicts. any updates on this one ? would really love to see it land.
let me know if i can do anything to help out

Manciukic commented on src/vmm/src/devices/virtio/mem/device.rs:
```diff
@@ -347,17 +353,17 @@ impl VirtioMem {
-        METRICS.plug_count.inc();
-        let _metric = METRICS.plug_agg.record_latency_metrics();
+        self.metrics.plug_count.inc();
+        self.metrics.plug_agg.record_latency_metrics();
```
**(High) Latency metric records ~0 here.** The return value of `record_latency_metrics()` is no longer bound, so the `LatencyMetricsRecorder` is dropped at the end of *this statement* instead of at the end of the operation. Its `Drop` impl (`logger/metrics.rs`) records `now - start_time`, so as written `start_time ≈ drop_time` and the metric is ~0. Only `mem` regressed — `net/device.rs` and `block/virtio/request.rs` still bind with `let _metric = …`. Same issue on lines 392, 415 and 441.

```suggestion
        let _metric = self.metrics.plug_agg.record_latency_metrics();
```

(Optionally marking `record_latency_metrics` `#[must_use]` would make this a compile error in future.)

aerosouund replied:
@Manciukic Great catch. Indeed i had missed this. But there's a small problem with simply bringing back the `_metric` use of the result. 
Other calls in that functions that use that latency recorder use mutable references to `self` (the virtio mem device). Meaning that we would be keeping alive an immutable reference and mutable ones which of course doesn't compile

```rust
        // immutable reference is created and is alive until the end of the function
        let _metric = self.metrics.plug_agg.record_latency_metrics(); 

        let response = match self.process_plug_request(range) { // mutable reference used
            Err(err) => {
                self.metrics.plug_fails.inc();
                error!("virtio-mem: Failed to plug range: {}", err);
                Response::error()
            }
            Ok(_) => {
                self.metrics
                    .plug_bytes
                    .add(usize_to_u64(self.nb_blocks_to_len(range.nb_blocks)));
                Response::ack()
            }
        };
```
As a workaround, i added a function of the recorder on the `LatencyMetricsRecorder` called `record_into` that you pass into it the start time and the metrics instance reference. Making the code look like this

```rust
        let start = get_time_us(ClockType::Monotonic);

        let response = match self.process_plug_request(range) {
            Err(err) => {
                self.metrics.plug_fails.inc();
                error!("virtio-mem: Failed to plug range: {}", err);
                Response::error()
            }
            Ok(_) => {
                self.metrics
                    .plug_bytes
                    .add(usize_to_u64(self.nb_blocks_to_len(range.nb_blocks)));
                Response::ack()
            }
        };
        let res = self.write_response(response, resp_addr, used_idx);
        LatencyMetricsRecorder::record_into(start, &self.metrics.plug_agg);
        res
```

Not the cleanest solution in the world. If you have better ideas let me know. I have kept the change in a single commit in case we decided to discard it. see the commit with message `refactor: use an associated function for latency recording`

Manciukic commented on src/vmm/src/devices/virtio/mem/device.rs:
```diff
@@ -382,16 +388,16 @@ impl VirtioMem {
-        METRICS.unplug_count.inc();
-        let _metric = METRICS.unplug_agg.record_latency_metrics();
+        self.metrics.unplug_count.inc();
+        self.metrics.unplug_agg.record_latency_metrics();
```
Same as line 357 — the recorder is dropped immediately and records ~0.

```suggestion
        let _metric = self.metrics.unplug_agg.record_latency_metrics();
```

aerosouund replied:
see comment on above thread

Manciukic commented on src/vmm/src/devices/virtio/mem/device.rs:
```diff
@@ -405,15 +411,15 @@ impl VirtioMem {
-        METRICS.unplug_all_count.inc();
-        let _metric = METRICS.unplug_all_agg.record_latency_metrics();
+        self.metrics.unplug_all_count.inc();
+        self.metrics.unplug_all_agg.record_latency_metrics();
```
Same issue here.

```suggestion
        let _metric = self.metrics.unplug_all_agg.record_latency_metrics();
```

aerosouund replied:
see comment on above thread

Manciukic commented on src/vmm/src/devices/virtio/mem/device.rs:
```diff
@@ -431,11 +437,11 @@ impl VirtioMem {
-        METRICS.state_count.inc();
-        let _metric = METRICS.state_agg.record_latency_metrics();
+        self.metrics.state_count.inc();
+        self.metrics.state_agg.record_latency_metrics();
```
Same issue here.

```suggestion
        let _metric = self.metrics.state_agg.record_latency_metrics();
```

Manciukic commented on src/vmm/src/devices/virtio/vsock/persist.rs:
```diff
@@ -109,7 +110,17 @@ where
+            // If we are here then the muxer must have been initiated. Get the metrics
+            // instance and copy it to pass it to the vsock struct.
+            let mut metrics_guard = METRICS.write().unwrap();
+            let metrics_inst = metrics_guard.remove(&state.cid).unwrap();
```
**(Medium)** `metrics_guard.remove(&state.cid).unwrap()` will panic during snapshot restore if there is no entry for `state.cid`, taking down the VMM on the host restore path (e.g. a construction-order change or an unexpected snapshot). `connection.rs` already uses the graceful `entry(...).or_insert_with(...)` pattern — consider the `Entry` API or `unwrap_or_else(|| Arc::new(VsockDeviceMetrics::default()))` here.

aerosouund replied:
ended up using `unwrap_or_else`. feel free to resolve thread

Manciukic commented on src/vmm/src/devices/virtio/vsock/unix/muxer.rs:
```diff
@@ -1474,49 +1478,36 @@ mod tests {
 
-        // Check that a new connection was added.
-        assert_eq!(METRICS.conns_added.count(), conns_added + 1);
+        assert_eq!(ctx.muxer.metrics.conns_added.count(), 1); // there is some bs going on here
```
**(Medium)** The `conns_removed` assertions are commented out (lines 1483 and 1529) and there's a `// there is some bs going on here` note on this `conns_added` check. Connection add/remove accounting is exactly what this refactor risks breaking, so it'd be good to port these assertions to the per-device model rather than disable them — or leave a `TODO` linked to a tracking issue if it needs to wait.

Manciukic commented on src/vmm/src/devices/virtio/rng/metrics.rs:
```diff
@@ -36,20 +36,26 @@
 
-/// Stores aggregated entropy metrics
-pub(super) static METRICS: EntropyDeviceMetrics = EntropyDeviceMetrics::new();
+use std::collections::BTreeMap;
```
**(Minor)** `use std::collections::BTreeMap;` is unused in this file (entropy uses `OnceLock`, not a map) and will fail under `deny(warnings)`. Please remove it.

Manciukic commented on src/vmm/src/devices/virtio/balloon/device.rs:
```diff
@@ -1696,10 +1697,11 @@ pub(crate) mod tests {
-                METRICS.free_page_hint_freed,
-                0,
-                self.th.device().process_free_page_hinting_queue()
+            let before = self.th.device().metrics.free_page_hint_freed.count();
```
**(Minor)** `let before = …` is never used afterward (the assertion checks the absolute `expected`). Leftover from the old before/after macro — unused-variable warning; please remove it.

Manciukic commented on src/vmm/src/devices/virtio/mem/metrics.rs:
```diff
@@ -24,20 +24,26 @@
 
 /// Called by METRICS.flush(), this function facilitates serialization of virtio-mem device metrics.
 pub fn flush_metrics<S: Serializer>(serializer: S) -> Result<S::Ok, S::Error> {
+    // why are even doing different handling ? what is seq and why is it important ?
```
**(Nit)** Leftover authoring note: `// why are even doing different handling ? what is seq and why is it important ?` — please drop before merge.

Manciukic commented on src/vmm/src/devices/virtio/rng/metrics.rs:
```diff
@@ -84,17 +90,18 @@ impl EntropyDeviceMetrics {
-        assert_eq!(entropy_metrics.entropy_event_count.count(), 1);
+    fn test_rng_dev_metrics() {
+        let metrics_instance = Arc::new(EntropyDeviceMetrics::new());
+        METRICS.set(metrics_instance);
```
**(Test — depends on global `METRICS`)** `test_rng_dev_metrics` drives the global `OnceLock` (`METRICS.set(...)` + `METRICS.get().unwrap()`), which isn't parallel-safe: whichever test initializes the `OnceLock` first wins the slot, so the exact `assert_eq!(host_rng_fails == 5)` on line 105 becomes order-dependent. This is the coupling the PR aims to remove — please assert on a local `EntropyDeviceMetrics` instance instead. (Also, this line ignores the `#[must_use]` `Result` from `OnceLock::set`; production uses `let _ = METRICS.set(...)`.)

Manciukic commented on src/vmm/src/devices/virtio/balloon/metrics.rs:
```diff
@@ -100,15 +105,21 @@ impl BalloonDeviceMetrics {
+        // its fine if .set() errors here. due to the METRICS lock being initialized
+        // by another test. the test flushes the global instance metrics and checks
+        // for structural similarity only.
+        METRICS.set(balloon_metrics.clone());
```
**(Test — depends on global `METRICS`)** `test_balloon_dev_metrics` uses the global `OnceLock`. Same parallelism concern as the rng test — please build and assert on a local `BalloonDeviceMetrics` instance.

Manciukic commented on src/vmm/src/devices/virtio/pmem/metrics.rs:
```diff
@@ -214,7 +188,12 @@ pub mod tests {
             let pmem_name: String = format!("pmem{}", i);
-            metrics.push(PmemMetricsPerDevice::alloc(pmem_name.clone()));
+            let dev_metrics = Arc::new(PmemMetrics::default());
+            METRICS
```
**(Test — depends on global `METRICS`)** `test_max_pmem_dev_metrics` mutates the shared global map directly (the `drop(METRICS.read()/write())` + `insert` dance is a sign it's fighting shared state). Please exercise a local `Arc<PmemMetrics>` / local map so the test is independent.

Manciukic commented on src/vmm/src/devices/virtio/net/metrics.rs:
```diff
@@ -280,27 +252,29 @@ pub mod tests {
             let devn: String = format!("eth{}", i);
-            NetMetricsPerDevice::alloc(devn.clone());
+
+            METRICS
```
**(Test — depends on global `METRICS`)** `test_max_net_dev_metrics` inserts into and reads from the global `METRICS` map. Please migrate to local instances (same applies to `test_signle_net_dev_metrics` below).

Manciukic commented on src/vmm/src/devices/virtio/net/metrics.rs:
```diff
@@ -353,13 +324,14 @@ pub mod tests {
 
-        NetMetricsPerDevice::alloc(String::from(devn));
-        METRICS.read().unwrap().metrics.get(devn).unwrap();
+        METRICS
```
**(Test — depends on global `METRICS`)** Same here — `test_signle_net_dev_metrics` depends on the global map; build a local instance instead.

Manciukic commented on src/vmm/src/devices/virtio/vsock/metrics.rs:
```diff
@@ -123,23 +131,253 @@ impl VsockDeviceMetrics {
+        let guest_cid: u64 = 10;
+
+        // Drop any existing read/write lock to avoid deadlocks or stale locks.
+        drop(METRICS.read().unwrap());
```
**(Test — depends on global `METRICS`)** `test_vsock_default` resets and mutates the global map (`drop(METRICS.read()/write())` then `insert(cid, …)`). You already have the target pattern in this same file — `test_vsock_metrics_serialization` (line 250) and `test_vsock_metrics_aggregation` (line 346) use a local instance with no global access. Please bring this test in line with those.

Manciukic commented on src/vmm/src/devices/virtio/vsock/csm/connection.rs:
```diff
@@ -509,6 +513,16 @@ where
+        // If a muxer already added a metrics instance to the map.
+        // otherwise if this connection is being initialized separately
+        // in a test let it create its own metrics instance
+        let metrics = METRICS
```
**(Test parallelism — core to #4709)** Connections resolve their metrics from the global CID-keyed map here via `entry(peer_cid).or_insert_with(...)` rather than receiving the owning device/muxer's `Arc`. Combined with the `insert(cid, …)` overwrite in `device.rs:115` / `muxer.rs:323`, two unit tests sharing a CID (the test constants `PEER_CID = 3`, `LOCAL_CID = 2`, and `test_utils.rs` `CID = 52`) can cross-wire: one test's connection can pick up another test's `Arc`, or have its map entry overwritten, so `ctx.muxer.metrics` assertions read the wrong counts under parallel execution.

aerosouund replied:
Right. to resolve this i made the connection take an externally supplied metrics instance along with ensuring that an instance is supplied by always creating one in `TestBackend::new`. this way during tests, they won't hit the case where they need to create their own metrics instance.

We can also make this argument an Arc directly, not an option. Which would mean the caller to a connection instantiation function MUST supply a metrics instance.
See the change in `refactor: make the vsock TestBackend own metrics creation` and let me know what you think!

Manciukic reviewed (commented):
Thanks for taking this on, and for the clean split into per-device commits — it made review much easier. The direction is solid and the per-device model reads well. I went through it commit-by-commit and over the whole diff. A few things to sort before it's mergeable; none are deep, but one is a real functional regression.

Also, please run the following validation locally to check the state of the PR:
```
./tools/devtools checkbuild --all
/tools/devtool checkstyle
./tools/devtool test -- integration_tests/build/test_unittests.py

# alternatively to run single unit tests in parallel
./tools/devtool sh "cargo test -p vmm devices::virtio::vsock::metrics"
# or serially
./tools/devtool sh "cargo test -p vmm devices::virtio::net::metrics -- --test-threads=1"
```

aerosouund committed:
refactor: use an associated function for latency recording

- the latency recorder created through a call to record_latency_metrics
  needs an immutable reference to the metrics instance it will write
  back to. This previously wasn't an issue because the metrics instance
  lived in the global METRICS variable but now metrics have moved to be
  inside the virtio mem device. meaning that referencing it requires
  referencing the device.
- create an associated function on the LatencyMetricsRecorder that isn't
  bound to a particular instance called record_into, which takes the
  start time and the metric instance to write into.
- replace callers of record_latency_metrics to use this pattern.

Signed-off-by: aerosouund <aerosound161@gmail.com>

aerosouund committed:
refactor: make the vsock TestBackend own metrics creation

- the muxer is the owner of metrics instantiation in vsock, make the
  TestBacked (muxer equivalent in testing) create VsockDeviceMetrics arc
  during calls to TestBackend::new() so other entities during tests
  (connections and the vsock device) can use that backend's metrics
  instance.
- make the connection builders take an Option<Arc<VsockDeviceMetrics>>
  as an argument to delegate metrics creation to their caller. Only
  create metrics instances if that caller passed None.

Signed-off-by: aerosouund <aerosound161@gmail.com>

aerosouund commented:
@Manciukic 
Thanks for taking the time to review this.
I went through your comments. Here are the important changes since your review:
- the change mentioned in [this comment](https://github.com/firecracker-microvm/firecracker/pull/5599#discussion_r3429216137)
- deleted the `test_max_pmem_dev_metrics` which, if we had removed reliance on the global instance, becomes the same as the single instance test but with a vector. no new functionality covered
- made the vsock csm connection builders (`new_local_init` and `new_peer_init`) take an externally supplied metrics instance to resolve what you brought up in [this thread](https://github.com/firecracker-microvm/firecracker/pull/5599#discussion_r3420508410)
- removed the part in `test_balloon_dev_metrics` where we serialize the local metrics instance, flush the global one, and compare the two which was doing structural comparison between the two. did this to remove reliance on the global instance in tests as mentioned. We can maybe add a todo to add a test that checks the structure of the empty serialized balloon device metrics instance and ensures its adherence to what we want. Let me know if you agree.

I ran the checks you mentioned on the PR and they pass (same for checkbuild and checkstyle)
```
integration_tests/build/test_unittests.py::test_unittests PASSED                                                                     [ 50%]
integration_tests/build/test_unittests.py::test_benchmarks_compile PASSED                                                            [100%]
```

aerosouund commented:
@Manciukic 
Ping, let me know whenever you got a chance to check this

Manciukic commented on src/vmm/src/devices/virtio/vsock/metrics.rs:
```diff
@@ -39,21 +39,29 @@
+    let vsock_metrics = METRICS.read().unwrap();
+    let mut vsock_aggregated: VsockDeviceMetrics = VsockDeviceMetrics::default();
+    for (_, metrics) in vsock_metrics.iter() {
+        vsock_aggregated.aggregate(metrics);
```
**vsock metrics are no longer reset on flush, so they are reported as cumulative totals.**

`aggregate()` reads each instance with `fetch_diff()` (`count - baseline`), but the baseline is only advanced by `SharedIncMetric::serialize` (`logger/metrics.rs:279`). Since `8525f4ad6` we serialize only `vsock_aggregated` (line 56) and never the per-CID instances, so their baselines never move and every flush re-adds the full count. Guest sends one packet per interval: `main` emits `tx_packets_count` 1, 1, 1; this emits 1, 2, 3... `net/metrics.rs` still does `serialize_entry(&devn, m)` per device, which is why it is unaffected. Schema validation in `fcmetrics.py` will not catch this.

vsock is single-instance, so the clean fix is to drop the map:

```rust
pub(super) static METRICS: RwLock<Option<Arc<VsockDeviceMetrics>>> = RwLock::new(None);
```

and serialize the instance directly (`Some(m) => seq.serialize_entry("vsock", m)?`, `None => ...default()`). That restores the reset semantics, keeps the JSON identical, and lets you delete the CID key, the `Option<Arc<_>>` parameters on `Vsock::new`/`with_queues` and both connection constructors, the `or_insert_with` fallback, and the `remove`/`insert` dance in `persist.rs:113-123`.

aerosouund replied:
Fixed by removing the per CID map. I realized its an invalid change in the first place since there can be only one vsock device per guest as of my current understanding. As well as not using an option of arc in the arguments and pass an arc directly

Manciukic commented on src/vmm/src/devices/virtio/mem/device.rs:
```diff
@@ -432,11 +447,11 @@ impl VirtioMem {
-        METRICS.state_count.inc();
-        let _metric = METRICS.state_agg.record_latency_metrics();
+        self.metrics.state_count.inc();
+        self.metrics.state_agg.record_latency_metrics();
```
**`state_agg` still records ~0 here** — this one recorder never got converted, so the guard is dropped at the `;`.

Rather than fixing just this line, I would go back to the RAII guard everywhere and sidestep the borrow conflict by cloning the `Arc` first. Your diagnosis was right: `record_latency_metrics()` borrows `self.metrics`, so the guard cannot live across the `&mut self` calls. But the guard only needs to borrow *an* `Arc`, not `self`'s:

```rust
    fn handle_plug_request(
        &mut self,
        range: &RequestedRange,
        resp_addr: GuestAddress,
        used_idx: u16,
    ) -> Result<(), VirtioMemError> {
        // Clone keeps the recorder's borrow off `self` so it can outlive the calls below.
        let metrics = Arc::clone(&self.metrics);
        metrics.plug_count.inc();
        let _metric = metrics.plug_agg.record_latency_metrics();

        let response = match self.process_plug_request(range) {
            /* unchanged */
        };
        self.write_response(response, resp_addr, used_idx)
    }
```

`_metric` is declared after `metrics`, so it drops first (the compiler enforces that ordering) and the guard then covers `write_response` and any `?` early return. Cost is one refcount bump per plug/unplug/state request, negligible on a hotplug path.

That also lets you revert `33caec6f3` entirely: `LatencyMetricsRecorder::record_into` and its `logger/mod.rs` export go away, the duplicated min/max/sum body goes with them, and this PR stops touching the logger. Remember to drop the then-unused `utils::time::{ClockType, get_time_us}` import from this file. Same shape for all four handlers: 352, 388, 414 and 444.

aerosouund replied:
Neat suggestion. Applied the arc cloning, dropped the commit that introduces `record_into` and fixed the `handle_plug_request` function to make it have the same pattern

Manciukic commented on src/vmm/src/devices/virtio/balloon/metrics.rs:
```diff
@@ -35,22 +35,27 @@
 
 /// Stores aggregated balloon metrics
-pub(super) static METRICS: BalloonDeviceMetrics = BalloonDeviceMetrics::new();
+pub(super) static METRICS: OnceLock<Arc<BalloonDeviceMetrics>> = OnceLock::new();
```
**`OnceLock` is first-writer-wins, which silently loses metrics for a re-configured device.**

I think this was a miss on my side initially. I didn't realise that despite only having a single instance, the object can get rebuilt from the API (`PUT /balloon` or `PUT /entropy`). Using a `RwLock` would be simpler to keep a `last-writer-wins` semantics. We can update all devices using `OnceLock` to `RwLock` with little changes.

```rust
pub(super) static METRICS: RwLock<Option<Arc<BalloonDeviceMetrics>>> = RwLock::new(None);
// in Balloon::new / from_state - last device wins, matching the builder's overwrite semantics
*METRICS.write().unwrap() = Some(metrics.clone());
```

The `warn!` then goes away. Same for `rng/metrics.rs:54` + `rng/device.rs:85` and `mem/metrics.rs:32` + `mem/device.rs:160`. virtio-mem is only built once (`builder.rs:634`), so there it is the consistency argument rather than a live bug.

aerosouund replied:
Dropped the `OnceLock` in favor of an `RwLock` in the ballon, rng and mem devices

Manciukic commented on src/vmm/src/devices/virtio/rng/device.rs:
```diff
@@ -3,6 +3,7 @@
 
 use std::io;
 use std::ops::Deref;
+use std::rc::Rc;
```
**`cargo clippy --all --all-targets -- -D warnings` (run by `devtool checkbuild`) should be red on this head.** 

- `rng/device.rs:6` — `use std::rc::Rc;` unused
- `rng/device.rs:25` — `info` unused in the logger import
- `vsock/event_handler.rs:36` — `use ...vsock::metrics::METRICS;` unused
- `balloon/mod.rs:15` — `METRICS` unused in the import list
- `vsock/csm/connection.rs:98` — `use crate::vmm_config::vsock::VsockDeviceConfig;` unused
- `vsock/metrics.rs:107` — private `const fn new()` now has no callers (dead_code); the comment above it about `METRICS` not accepting `default()` is also obsolete
- `balloon/metrics.rs:108`, `rng/metrics.rs:91`, `pmem/metrics.rs:161` — `use std::sync::Arc;` unused in the test modules
- `block/virtio/metrics.rs:235` and `:266` — `devn` bindings unused since `alloc()` went away
- `net/metrics.rs:244` — loop variable `i` unused

While you are in there: `rng/device.rs:98` can move `metrics` instead of cloning it, and `vsock/metrics.rs:24` still documents "there is no per device metrics".

aerosouund replied:
Fixed by running the clippy command and going through the failures. I left the `vsock/metrics.rs:24` comment because the single device test was as you said.. tautological. There are now tests that cover the required functionality. Let me know if there is any thing still uncovered!

Manciukic commented on src/vmm/src/devices/virtio/net/metrics.rs:
```diff
@@ -267,164 +239,36 @@ pub mod tests {
-        drop(METRICS.write().unwrap());
+        let mut metric_instances = Vec::new();
 
         for i in 0..MAX_NET_DEVICES {
```
**The converted metrics tests can no longer fail, and nothing covers the registry any more.**

Each device now gets a fresh local instance, so `assert!(inst.activate_fails.count() >= 1)` and `>= 10` are tautological, and the comments about "another unit test running in parallel which updates drv0" no longer describe anything. Same in `test_single_net_dev_metrics:269`, `block/virtio/metrics.rs:228,262`, `pmem/metrics.rs:165` (whose `<= 2` bound and comment refer to the deleted `test_max_pmem_dev_metrics`), `balloon/metrics.rs:109` and `rng/metrics.rs:92`.

Net result: no test exercises the map registry or `flush_metrics` aggregation, and balloon/rng lost their serialization checks. Could you make the assertions exact (`== 1`, `== 10`), drop the stale comments, and keep one test per device that (a) serializes a `default()` instance and asserts the JSON shape and (b) aggregates over a locally built map? `test_vsock_metrics_aggregation` is already the model.

aerosouund replied:
Correct observation. Replaced with a testing scenario where we:
- Single instance device (mem, rng, vsock... etc): create an instance and change the counts of metrics inside it, serialize it, turn it to a json object and ensure that the counts on the object are the counts we modified on the rust type
- Multi instance device (net, pmem... etc): create multiple instances and an aggregator and aggregate them to the aggregator. ensure its counts are the sum of the individual instances, and that after serialization counts are equal to that sum as well.

Let me know if this is satisfactory

Manciukic commented on src/vmm/src/devices/virtio/balloon/mod.rs:
```diff
@@ -114,10 +114,13 @@ pub enum RemoveRegionError {
 }
 
-pub(super) fn report_balloon_event_fail(err: BalloonError) {
+pub(super) fn report_balloon_event_fail(
```
**Two copies of this function, and the `Arc` is taken by value.**

`Balloon::report_balloon_event_fail` was added at `device.rs:898` but nothing calls it (dead_code), while this free function is still the one used. Please keep one. Whichever survives should take `&BalloonDeviceMetrics` rather than an owned `Arc` — right now every error path in `event_handler.rs:121-136` does a `.clone()` (six sites) purely to satisfy the signature.

aerosouund replied:
settled for the free function and deleted the method inside the `impl`. removed the clone by making the argument an immutable reference

Manciukic commented on src/vmm/src/logger/metrics.rs:
```diff
@@ -807,6 +822,7 @@ impl LatencyAggregateMetrics {
     /// 2nd for delta_time_us = get_time_us() - start_time; and metrics.store(delta_time_us)
     /// we have just `_m = metrics.record_latency_metrics()`
+    #[must_use]
     pub fn record_latency_metrics(&self) -> LatencyMetricsRecorder<'_> {
```
`record_into` at `:767` can go away entirely if you take the `Arc`-clone approach from my `mem/device.rs` comment.

aerosouund replied:
Removed it by following your suggestion of arc cloning

Manciukic commented on src/vmm/src/devices/virtio/balloon/device.rs:
```diff
@@ -1800,12 +1803,12 @@ pub(crate) mod tests {
         // so we should ignore any inflight requests
         ht.start_hinting(None);
-        ht.test_hinting(None, 0);
+        ht.test_hinting(None, 4096);
```
Hardcoded page size here (`4096`) and at `:1853` (`8192`), while every neighbouring call uses `ht.page_size` from `host_page_size()`. Please use `ht.page_size` / `ht.page_size * 2` so the test still holds on hosts with non-4K pages.

aerosouund replied:
Dropped the absolute values in the favor of `ht.page_size`

Manciukic commented on src/vmm/src/devices/virtio/block/virtio/metrics.rs:
```diff
@@ -86,50 +86,23 @@ use serde::{Serialize, Serializer};
-static METRICS: RwLock<BlockMetricsPerDevice> = RwLock::new(BlockMetricsPerDevice {
-    metrics: BTreeMap::new(),
-});
+pub static METRICS: RwLock<BTreeMap<String, Arc<BlockDeviceMetrics>>> =
```
These registries were private statics before; `pub` under `pub mod devices` makes them part of the crate's public API. `pub(crate)` is enough — same for `pmem/metrics.rs:92`, `net/metrics.rs:90` and `vsock/metrics.rs:59`.

aerosouund replied:
Done

aerosouund commented:
@Manciukic 
Thank you for the thorough review.
I have addressed your comments, please check the individual responses on the threads you started and ping me if anything can still be made better.

Apart from the threads i have:
- Applied the same pattern to the block vhost user device, had missed it originally
- I have squashed all commits that contain test fixes into the original device commits
- Used absolute value assertion and included that in the original device commits
- Ensured `checkbuild`/`checkstyle` pass on the current state of the code after rebasing on main

Regarding msix/mmds logger globals, i haven't taken a look at them in the context of this PR. I can do that if you'd like, or in a follow up which would also include the change of `RUST_TEST_THREADS` to be more than one
